python/lsst/pipe/supertask/configOverrides.py

```diff
@@ -81,6 +81,11 @@ def applyTo(self, config):
                 obj = config
                 for attr in field[:-1]:
                     obj = getattr(obj, attr)
+                # Check if a list was passed as denoted by enclosing brackets, and exec the list
+                # must be done before the try, as it will set a string which is a valid iterable
+                # object and will be assigned instead of the list
+                if isinstance(value, str) and value.startswith('[') and value.endswith(']'):
+                    value = eval(value)
```

Comment:
Hmm, that's too easy (and it opened nice `eval` backdoor) :) What if I want to pass for example string `[0-9a-zA-Z]` to a string field?

I believe this code was borrowed (with modification) from CmdLineTask, and my impression was that cases like this are supposed to be handled in `except` clause (but apparently that did not work for you). I actually hate this code because of `eval` and unreliable error detection but I do not know how to fix it. Maybe we should ask for help from `pex_config` experts?

In addition to fix itself we'd also need a unit test for it.

Reply:
So I too am a bit bothered by the eval, and a longer term solution will probably involve a custom parser ( which I have one we can use actually), but this gets things up and running until such a time. It is itself no more dangerous that what we already have with the eval, or how we load and run config files that could contain arbitrary code.

As far as your concern that someone may want to pass a regex expression (or some other similar string) I would argue that is the minority if we have it at all, and should be special cased instead of lists, which are used all over the place.

The reason that this will not get executed in the try statement is that in order to put a list in on the command line, it must be encased in quotes, assuring bash that this is a string. This is normal and fine. However, in python a string is a valid iterable. This means that the try block happily took the string as a list, and attempted to iterate over each of the characters, which of course did not work.

I agree that a more principled approach  should be done in the future. The work of input sanitation however is very nuanced, and will require a developer a decent amount of time to get right. This code will allow us to use these configuration options in the mean time, at no more risk than we already were in, and at the trade off of obscure string configs. I would rather need to put obscure string config in a config file to load them as overrides than every instance of a list in the config system.

Reply:
This particular `eval` can be made slightly safer by passing empty dict as second argument as we do in other places. I wonder if we can use [ast.literal_eval](https://docs.python.org/3.6/library/ast.html?highlight=ast#ast.literal_eval) here, it is limited in that it does not support expressions but it supports lists of constants.

I did not say `[0-9a-zA-Z]` is a regular expression 🙂 My point is that someone may pass a string that starts and ends with square brackets to a string field (or any other non-list field). `eval` should only be called for ListField but not for other types of fields (well, maybe it should be called for some types, but not all). Ideally that sort of logic belongs to `pex_config` iself, maybe we need `Field.setFromString()` method to replace this code which is very fragile.

Reply:
I agree this is not the most principled or stable feature, but that is not the change-set I was trying to introduce. To properly handle loading and overriding configs is going to take significant developer time and thought, or at least more than a side note while working on other things. I am simply trying to expand the scope of existing config overrides that exist now. This code is certainly not worse off than what we have now.

As indicated above , "(or some other similar string)", I know you were not limiting yourself to regular expressions, I was merely using that as a likely string to contain such character arrangements for the purpose of the discussion.

I am not approaching this change set from a computer scientist, best practices, mindset as I would if I was trying to bring a final solution to this code block. I am coming at this as a developer familiar with the stack, how it is used day to day, and what options we have. The scenario where a string containing those characters must be set is so unlikely I would be surprised if there was on the order of one case in our code base, as opposed to the tens (many tens?) of cases where a list field is used.

We can reject this change-set (or something very similar) on the basis that it might in some scenario conflict with a string someone wants to enter, but that would leave us in the situation we currently have where we cannot set any list field from the command line. List fields must be set in a file and then loaded in. I would rather that be something done on the off chance someone needs that sort of string, as it will be much less often.

If you or someone else has time in their schedules to do something more robust and thought out, great! However I think it would be a shame to let the perfect get in the way of the good enough for now, if no one can get to it for weeks, or months (or years as list configs have never been able to be set).

I certainly have no problem using ast.literal_eval in this case, it would be a bit safer, not that our code is currently safe. It would not however address the concern about a string input with brackets. It could be wrapped in a try except block to handle some of those cases, but it is not a sure thing that a pathological string would not caught up by the system. For instance it would catch "[A-Z]", but not "[dir]".

Reply:
My opinion: the `eval` should definitely have a `{}` argument, and the test should not be primarily on the value string syntax but instead on the `obj` type (check if it's an instance of `lsst.pex.config.ListField`, then eval if the value is a string).  That would also allow `ListField`s to be overridden with `()` tuples in addition to `[]` lists, I believe.